File: modules/news_app/src/ios/ValdiWebView.m
// ValdiWebView.m
#import "ValdiWebView.h"

// Declare the Valdi binding context method to satisfy the compiler
@interface NSObject (ValdiAttributeBinding)
- (void)bindAttribute:(NSString *)name
invalidateLayoutOnChange:(BOOL)invalidate
      withStringBlock:(void (^)(id view, NSString *value, id animator))applyBlock
           resetBlock:(void (^)(id view, id animator))resetBlock;
@end

@interface ValdiWebView ()
@property (nonatomic, strong) NSString *urlString;
@end

@implementation ValdiWebView

- (instancetype)initWithFrame:(CGRect)frame configuration:(WKWebViewConfiguration *)configuration {
    self = [super initWithFrame:frame configuration:configuration];
    if (self) {
        [self setupWebView];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)coder {
    self = [super initWithCoder:coder];
    if (self) {
        [self setupWebView];
    }
    return self;
}

- (void)setupWebView {
#if TARGET_OS_IPHONE
    // Enable scrolling (iOS only)
    self.scrollView.scrollEnabled = YES;
    self.scrollView.bounces = YES;
    
    // Ensure links are clickable
    self.userInteractionEnabled = YES;
    
    // Disable gesture recognizers that might interfere
    for (UIGestureRecognizer *gesture in self.gestureRecognizers) {
        gesture.delaysTouchesBegan = NO;
        gesture.delaysTouchesEnded = NO;
    }
#endif
    
    // Allow all types of interaction (works on both iOS and macOS)
    self.allowsBackForwardNavigationGestures = YES;
}

- (void)valdi_applyURL:(NSString *)url {
    self.urlString = url;
    if (url && url.length > 0) {
        NSURL *nsUrl = [NSURL URLWithString:url];
        if (nsUrl) {
            NSURLRequest *request = [NSURLRequest requestWithURL:nsUrl];
            [self loadRequest:request];
        }
    }
}

+ (void)bindAttributes:(id)bindingContext {
    [bindingContext bindAttribute:@"url"
         invalidateLayoutOnChange:NO
                  withStringBlock:^(ValdiWebView *view, NSString *url, id animator) {
        [view valdi_applyURL:url];
    }
                       resetBlock:^(ValdiWebView *view, id animator) {
        [view valdi_applyURL:nil];
    }];
}

@end

File: modules/news_app/src/ios/ValdiWebView.swift
// ValdiWebView.swift
import WebKit
import Foundation

@objc class ValdiWebView: WKWebView {
    @objc var urlString: String? {
        didSet {
            if let urlString = urlString, let url = URL(string: urlString) {
                self.load(URLRequest(url: url))
            }
        }
    }
    
    override init(frame: CGRect, configuration: WKWebViewConfiguration) {
        super.init(frame: frame, configuration: configuration)
        setupWebView()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        setupWebView()
    }
    
    private func setupWebView() {
        // Enable scrolling
        self.scrollView.isScrollEnabled = true
        self.scrollView.bounces = true
        
        // Allow all types of interaction
        self.allowsBackForwardNavigationGestures = true
        
        // Ensure links are clickable
        self.isUserInteractionEnabled = true
        
        // Disable gesture recognizers that might interfere
        for gesture in self.gestureRecognizers ?? [] {
            gesture.delaysTouchesBegan = false
            gesture.delaysTouchesEnded = false
        }
    }
    
    override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
        // Log hit testing for debugging
        NSLog("ValdiWebView hitTest at: \(point)")
        return super.hitTest(point, with: event)
    }
    
    // Bind the url attribute to Valdi
    @objc static func bindAttributes(_ bindingContext: SCValdiAttributesBindingContext) {
        bindingContext.bindAttribute(
            "url",
            invalidateLayoutOnChange: false,
            withStringBlock: { (view: ValdiWebView, url: String, animator: SCValdiAnimator?) in
                view.urlString = url
            },
            resetBlock: { (view: ValdiWebView, animator: SCValdiAnimator?) in
                view.urlString = nil
            }
        )
    }
}